src/nvim/lua/treesitter.c

```diff
@@ -1046,8 +1081,11 @@ static int query_next_capture(lua_State *L)
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
     if (n_pred > 0 && capture_index == 0) {
-      lua_pushvalue(L, lua_upvalueindex(4));  // [index, node, match]
```

Comment:
this was to avoid the creation of lua objects in the hot path. do we actually need this?

Reply:
Yes, because we need to clear the previous match values.
If you look at what changed in the tests, we actually did it wrong for a ling time, and we were keeping references in the temporary match to older nodes that are not in this match.

Reply:
I don't think so. For the changed test the old c code should work with the corrected match. Errors like the incorrect predicate can only be ruled out completely by static checking as predicate indicies can overlap anyway.

Reply:
No, the issue is that a capture used in the predicates was not actually part of the match, which is a problem.

Reply:
Ok, this was not wrong in the past, but is required now due to quantifiers possibly having zero nodes for a capture.

Reply:
it _was_ wrong in the past, one can _not_ use a node captured in the previous pattern to compute things in another pattern. Captures are specific to their pattern, and the bug that was previously there broke that, because one could use, in the current pattern, the captures from another pattern

Reply:
Sorry but this simply is not correct. In the past, before supporting quantifiers on captures, a match was always guaranteed to fill all of its captures (and do so exactly once). Therefore, any pattern with well-formed predicates would only use its own captures already, as no stale values could've been left behind. For a non-well-formed predicate that accesses an_ invalid_ capture there is _no_ correct value, such a query can and should be rejected at query compile time and is not a concern for the runtime hot loop.

But I am fine with de-optimizing this for now as it makes implementing quantifiers much simpler. We can consider optimizing again when we have proper profiling set up, anyway.